### PR TITLE
Added fix for AST module in order to be synchronized with the actual implementation

### PR DESCRIPTION
**Pull request by dorianverna17**
This PR is meant to synchronize the implementation of the Module class in astbase.d file with the implementation of the original Module class.

### CODE REVIEW

**Review by MoonlightSentinel — DISMISSED** [review]
Why does this revert several recent changes?
What is fixed?
What D-Scanner error?

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -5419,25 +5425,6 @@ struct ASTBase
         }
     }
 
-    extern (C++) final class ThrowExp : UnaExp
```
Comment (maxhaton) [inline src/dmd/astbase.d:5422]: ?

**Review by 12345swordy — CHANGES_REQUESTED** [review]
Where is the bugzilla report?

**Comment by RazvanN7** [comment]
@dorianverna17 concerns have been expressed regarding this PR. Please clarify what is the motivation/purpose of this PR.

**Comment by BorisCarvajal** [comment]
The issue/commit titles are really confusing.

But without the unrelated changes it's more clear the intent, you are changing some signatures to be more in line with the actual compiler classes (or actually to have a `Loc` info for better D-scanner diagnostics?).

I also wanted to change `ASTBase.Module` constructor to take a string array as in the real `Module` class, which makes more sense. 
Still, I recommend you to keep the old signature as a wrapper over the new one, there are some examples in `src/test/dub_package/` using it.

File: src/dmd/astbase.d [hunk 2]
```diff
@@ -1331,13 +1335,13 @@ struct ASTBase
...
 
-        extern (D) this(const(char)* filename, Identifier ident, int doDocComment, int doHdrGen)
+        extern (D) this(const ref Loc loc, const(char)[] filename, Identifier ident, int doDocComment, int doHdrGen)
         {
-            super(ident);
+            super(loc, ident);
             this.arg = filename;
-            srcfile = FileName(FileName.defaultExt(filename.toDString, mars_ext));
+            srcfile = FileName(filename);
         }
```
Comment (BorisCarvajal) [inline src/dmd/astbase.d:1346]: Keep the old signature.

```suggestion

extern (D) this(const(char)* filename, Identifier ident, int doDocComment, int doHdrGen)
{
    this(Loc.initial, filename.toDString, ident, doDocComment, doHdrGen);
}
```

File: src/dmd/astbase.d (hunk 1 above)
Comment (dorianverna17) [reply src/dmd/astbase.d]: I modified the PR to not include this change

**Comment by dorianverna17** [comment]
I updated the description of the PR as requested above and also made commits for the changes requested in the reviews above.

File: src/dmd/astbase.d (hunk 2 above)
Comment (RazvanN7) [reply src/dmd/astbase.d:1346]: ping @dorianverna17

Comment (dorianverna17) [reply src/dmd/astbase.d:1346]: I also kept the old signature in the commit made yesterday

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**